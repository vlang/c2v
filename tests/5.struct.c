#include <stdio.h>

const int PI= 314;

struct User {
	char *name;
	int age;
};

// lots of structs are typedef'ed in C
typedef struct {
	char *name;
	int age;
} TUser;

// struct names have to be capitalized in V
struct small {
	int foo;
};

struct small3 {
	int foo;
};

struct small2 {
	struct small struct_field; // make sure field types are capitalized
	struct small* struct_field_ptr;
	struct small3** struct_field_ptr2;
};

void small_fn(struct small param) { // make sure arg types are capitalized

}

typedef unsigned int angle_t;

enum Color {
	red, green, blue,
};

void handle_user(struct User user) {

}

void handle_tuser(TUser user) {

}

void multi_assign() {
	int aa = 0;
	int bb = 10;
	int cc = 20;
	aa = bb = cc;
}

void x(int pi) {}

#define arrlen(array) (sizeof(array) / sizeof(*array))

static int *weapon_keys[] = { 0, 0, 0 };
void sizeof_array() {
	int x = 10;
	int c = sizeof(x);
	int n = arrlen(weapon_keys);
}

int checkcoord[12][4] =
{
    {3,0,2,1},
    {3,0,2,0},
    {3,1,2,0},
    {0,0,0,0},
    {2,0,2,1},
    {0,0,0,0},
    {3,1,3,0},
    {0,0,0,0},
    {2,0,3,1},
    {2,1,3,1},
    {2,1,3,0}
};

void i_error(char* s, ...) {
	puts(s);
}

static unsigned long long sixtyfour(void) {
	return 64;
}

typedef union {
	int a;
	char b;
} MyUnion;

int main() {
	struct User user;
	user.age = 20;
	user.name = "Bob";
	printf("age=%d", user.age);
	handle_user(user);

	unsigned long long sf = sixtyfour();
	printf("sixtyfour=%lld", sf);
	// TODO
	// struct User user2 = { .age = 30, .name = "Peter" };
	TUser user2;
	user2.age = 30;
	user2.name = "Peter";
	handle_tuser(user2);
	//
	struct small s; // make sure struct inits are capitalized
	struct small s2 = { .foo = 10 };
	x(PI); // make sure cap consts stay capitalized
	return 0;
}
