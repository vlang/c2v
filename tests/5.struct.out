@[translated]
module main

@[export: 'PI']
const PI = 314

struct User {
	name &i8
	age  int
}

// lots of structs are typedef'ed in C
struct TUser {
	name             &i8
	age              int
	capitalizedField f32
	// fields in V must be lower case
}

// struct names have to be capitalized in V
struct Small {
	foo int
}

struct Small3 {
	foo int
}

struct Small2 {
	struct_field Small
	// make sure field types are capitalized
	struct_field_ptr  &Small
	struct_field_ptr2 &&Small3
}

fn small_fn(param Small) {
	// make sure arg types are capitalized
}

type Angle_t = u32

enum Color {
	red
	green
	blue
}

fn handle_user(user User) {
}

fn handle_tuser(user TUser) {
}

fn multi_assign() {
	aa := 0
	bb := 10
	cc := 20
	aa = cc
	bb = aa
}

fn x(pi int) {
}

@[export: 'weapon_keys']
const weapon_keys = [0, 0, 0]!

fn sizeof_array() {
	x := 10
	c := sizeof(x)
	n := (sizeof(weapon_keys) / sizeof(*weapon_keys))
}

@[export: 'checkcoord']
const checkcoord = [[3, 0, 2, 1]!, [3, 0, 2, 0]!, [3, 1, 2, 0]!,
	[0, 0, 0, 0]!, [2, 0, 2, 1]!, [0, 0, 0, 0]!, [3, 1, 3, 0]!,
	[0, 0, 0, 0]!, [2, 0, 3, 1]!, [2, 1, 3, 1]!, [2, 1, 3, 0]!]!

@[c2v_variadic]
fn i_error(s ...&i8) {
	C.puts(s)
}

fn sixtyfour() i64 {
	return 64
}

union MyUnion {
	a int
	b i8
}

struct AnonStructTest {
	age int
	foo struct {
		bar int
		baz &i8
	}

	last_field f32
}

struct AnonStruct_105 {
	fff int
	sss &i8
}

@[weak]
__global global_state_using_anon_struct AnonStruct_105

fn main() {
	user := User{}
	user.age = 20
	user.name = c'Bob'
	C.printf(c'age=%d', user.age)
	handle_user(user)
	sf := sixtyfour()
	C.printf(c'sixtyfour=%lld', sf)
	// TODO
	// struct User user2 = { .age = 30, .name = "Peter" };
	user2 := TUser{}
	user2.age = 30
	user2.name = c'Peter'
	handle_tuser(user2)
	//
	s := Small{}
	// make sure struct inits are capitalized
	s2 := Small{
		foo: 10
	}

	x(PI)
	// make sure cap consts stay capitalized
	return
}
